src/adapter/properties.rs

```diff
@@ -61,6 +61,17 @@ pub(super) fn resolve_item_property<'a>(
     }
 }
 
+pub(super) fn resolve_module_property<'a>(
+    contexts: ContextIterator<'a, Vertex<'a>>,
+    property_name: &str,
+) -> ContextOutcomeIterator<'a, Vertex<'a>, FieldValue> {
+    match property_name {
+        "is_crate" => resolve_property_with(contexts, field_property!(as_module, is_crate)),
```

Comment:
This is an entirely reasonable design here. But if you're up for it, I think we can take advantage of an advanced little Trustfall trick and do something clever and more ergonomic.

Here's the key observation that generally takes some getting used to when starting to use Trustfall: when modelling a data source in Trustfall, we don't actually have to map the underlying data representation one-to-one. We can model the _semantic_ representation of the underlying data, instead of matching how it is _syntactically_ represented in the underlying data source.

Here's what I mean.

Consider when we'd use the `is_crate` property: it's when we want to find the top-level module of the crate. The query for that might look like something like this:
```graphql
{
    Crate {
        item {
            ... on Module {
                is_crate @filter(op: "=", value: ["$true"])
                
                < ... etc ... >
            }
        }
    }
}
```

This makes sense, and is a one-to-one match for the data representation in the underlying JSON file. But it's also a bit clunky, no?

Consider this alternative (names subject to bikeshedding):
```graphql
{
    Crate {
        root_module {
            # this is already of type Module and represents the crate root
            < ... same stuff as the "etc." in the earlier query ... >
        }
    }
}
```

Now, `root_module` isn't an explicit relationship that is ever named as such in the underlying representation — everything in the JSON file goes through the item index — but this is what the _actual, truly-underlying data_ looks like, semantically. Nothing is stopping us from implementing such an edge in Trustfall, and it's often both a query ergonomics improvement _and_ it gives us some more decoupling from the specific implementation choices of the JSON representation.

I know that was a lot to take in at once. Did that make sense? Happy to answer any questions, or even hop on a quick video call to chat about it, if that would be helpful.

Reply:
I think this is a fine idea, but I'm not 100% sure how to actually build this. I guess this would mean adding a "root_module" edge in the graphql file, and then adding a "root_module" handler in resolve_crate_items, but at that point I'm afraid I'm lost.  The best I can think to do, based on copy-and-paste, is something like:

```
	"root_module" =>
	    Box::new(
		optimizations::item_lookup::resolve_crate_items(adapter, contexts, resolve_info).filter(|item| {
                  // check if item is a module with is_crate == true
                })
	    ),
```

but I'm not really sure that's right.

Reply:
That would work just fine, well done!

We can do slightly better by taking advantage of the JSON format structure: the JSON file has a key called `"root"` at top level whose value is the item ID of the root module. We can look up that ID in the crate index directly, which will let us avoid iterating through all items in the crate.

Reply:
Okay. I tried this:
```
	"root_module" => resolve_neighbors_with(contexts, move |vertex| {
	    let current_crate = adapter.current_crate;
	    let previous_crate = adapter.previous_crate;

	    let origin = vertex.origin;
	    let item_index = match origin {
                Origin::CurrentCrate => &current_crate.inner.index,
                Origin::PreviousCrate => {
		    &previous_crate
                        .expect("no previous crate provided")
                        .inner
                        .index
                }
	    };

	    let module = item_index.get(&Id("root".into())).expect("crate had no root module");
	    Box::new(std::iter::once(origin.make_item_vertex(module)))
	}),
```

but I'm afraid I ran into lifetime issues.  Do I need to pass current_crate and previous_crate  as arguments into this function manually?

Reply:
This is a common issue: `adapter` itself doesn't necessarily live long enough to be moved in, but `adapter.current_crate` and `adapter.previous_crate` are `&'a` references so they live long enough. Assign them outside the closure and move them into it instead of moving `adapter` — almost all the other edge implementations elsewhere do the same thing for the same reason.

Reply:
Ah, that did it!

Reply:
Given the new edge, I'd remove this property since I think it won't be that useful — if someone wants the crate-level module, they can get it directly via the new edge.

Reply:
Hm. What if you wanted to get all of the modules _except_ for the root module? Would that justify this property?

Reply:
That's still doable and in a way that more closely maps to the semantic data model (the query says "not the root module" explicitly here) instead of mapping to the `is_crate` field which is mostly for JSON representation purposes, not something Rust modules inherently have.

Here's how:
```graphql
{
    Crate {
        root_module {
            id @tag
        }

        item {
            ... on Module {
                id @filter(op: "!=", value: ["%id"])
                
                < ... these are all modules that aren't the root ... >
            }
        }
    }
}
```